lib/psci/psci_on.c

```diff
@@ -114,6 +114,13 @@ int psci_cpu_on_start(u_register_t target_cpu,
 		assert(psci_get_aff_info_state_by_idx(target_idx) == AFF_STATE_ON_PENDING);
 	}
 
+	/*
+	 * Store the re-entry information for the non-secure world here
+	 * before the core that needs it is started, this way we will not
+	 * need a lock to prevent a race to for this data.
+	 */
+	cm_init_context_by_index(target_idx, ep);
+
```

Comment:
But this de-optimises the code by doing more work before the new core is started, which is not what we want generally. I can see you're doing this to remove spinlock usage but that's a separate issue...

Reply:
The comment states this lock is only used for local CPU coordination which makes spinlock safe. The use of being a sync point by another CPU was added later. The issue is spinlocks do not work in many cases across cores and clusters. Because of this all the non-cpu-pd locks are bakery locks (at least on non-HW-coherent platforms). So why then why is a spinlock suddenly okay here? My guess is it is not, I think most platforms get lucky and finish writing the context before the target CPU needs it so the lock is not used. If you for testing add a delay loop before context initialization then you will start seeing many other platforms start failing as the lock will fail and the target CPU will run past and read uninitialized data.

I am not the first to run into this: https://github.com/ARM-software/tf-issues/issues/451

Lets take the verifiable correct thing and just have the data ready for the target before releasing it, micro-optimizing the ordering is causing random failures for us and others.

Reply:
The reason bakery locks are used in some places (on non HW_ASSISTED_COHERENCY platforms) is that in those cases, the locks are released while caches are disabled and so it's not safe to use spinlocks. Here, the locks are acquired and released on both the running and target CPUs while caches are enabled, using a memory type that achitecturally guarantees that a global Exclusives monitor is implemented. Hence it is architecturally safe to use spinlocks.

However, it's possible that there is a bug, or that not all platforms implement the architecture correctly. If there's a bug, we should identify and fix it. If we need to help broken hardware, then we should add abstractions for that, rather than hitting performance for everyone.

> If you for testing add a delay loop before context initialization then you will start seeing many other platforms start failing as the lock will fail and the target CPU will run past and read uninitialized data.

Can you provide more information here? If there's a test case and a list of platforms for which this fails, that may help identify the root problem.

Reply:
Hmm, well let me go back and take another look at our locks and see if I can root cause the issues I'm seeing locally, then lets go from there. This PR can be closed for now, I'll re-submit a different fix depending on what I find.